internal/controllers/gateway/gateway_utils.go

```diff
@@ -265,13 +265,65 @@ func getListenerStatus(
 		if listener.Hostname != nil {
 			hostname = *listener.Hostname
 		}
+		supportedkinds, ResolvedRefsReason := getListenerSupportedRouteKinds(listener)
+
+		// If the listener uses TLS, we need to ensure that the gateway is granted to reference
+		// all the secrets it references
+		if listener.TLS != nil {
+			TLSresolvedRefReason := string(gatewayv1beta1.ListenerReasonResolvedRefs)
+			for _, certRef := range listener.TLS.CertificateRefs {
+				// if the certificate is in the same namespace of the gateway, no ReferenceGrant is needed
+				if certRef.Namespace != nil && *certRef.Namespace != (Namespace)(gateway.Namespace) {
+					// get the result of the certificate reference. If the returned reason is not successful, the loop
+					// must be broken because the secret reference isn't granted
+					TLSresolvedRefReason = getReferenceGrantConditionReason(gateway.Namespace, certRef, referenceGrants)
```

Comment:
not so related: since `ReferenceGrant` is promoted to `v1beta1`, I think we should enable it by default if feature gate `Gateway` is enabled. Now if `ReferenceGrant` is disabled by disabling `GatewayAlpha`(which is by default), we will always fail here if secret is in a different namespace.

Reply:
This is a very good point, I'll create an issue for that

Reply:
see PR: #3519 to enable `ReferenceGrant` by default.